### PR TITLE
State is passed explicitly to instrumentation and parameters are NOT mutated

### PR DESCRIPTION
**Pull request by bbakerman**
As per https://github.com/graphql-java/graphql-java/discussions/2743 - object allocation at scale can be problematic because in a chained instrumentation call we create new `instrumentation parameters` per chained call  and we create a chained `InstrumentationContext` object (list of all the other `InstrumentationContext` objects) which gets held in memory for a long time and is not eligible for GC.

This PR starts to change `Instrumentation` so that

1. The `InstrumentationState` is passed into each method direct (no need for new objects to be allocated)
2. The returned `InstrumentationContext` can be null and hence no object allocation is needed here either

All the places that received a `InstrumentationContext` have been changed to swap in a static NooP object (singleton) and hence they dont change a lot in a consuming sense.

This PR as is shows that ALL the tests pass - that is the new methods are delegated back to the old no problems.

Some important decisions I made

* I used `null` and not Optional for the empty `InstrumentationContext` case - why?  Because there is no object allocation involved in the `some` case versus the empty case.

* The `instrumentXXX` methods have been changed to take in `InstrumentationState` for consistency reasons with all the others even if the memory pressure problem doesnt apply to them as much
* The `instrumentXXX` methods MUST return a non null object as before - I put `@Nonnull` annotations to this better

The ChainedInstrumentation class has been updated to copy with null contexts being returned returned AND it only uses the new methods with state and never the old.

Still TODO

I have not updated ALL the provided instrumentations to take advantage of the new methods.  I think this is best dont in a another PR to reduce the scope of this already large PR.  They all work unchanged so lets not make it more complicated than it needs to be and we can improve it via another PR

### CODE REVIEW

**Comment by jonclarkethg** [comment]
Hi. Wondering if there is an update on this as we're very keen to see the impact on object allocation. Thanks :)

**Comment by bbakerman** [comment]
I am waiting on a final review from @andimarek - Andi?

**Comment by jonclarkethg** [comment]
@andimarek did you get a chance to review this PR?

**Comment by Aynemayne** [comment]
Hello @andimarek and/or @bbakerman! Did you manage to follow up on this PR? 
Thanks in advance!

**Comment by bbakerman** [comment]
I just fixed up the merge conflict on this PR - this will be merged into 19.0 - 19.0 is to be released very soon - probably next week

**Review by andimarek — APPROVED** [review]

File: src/main/java/graphql/execution/instrumentation/ChainedInstrumentation.java
Comment (josephlbarnett) [inline src/main/java/graphql/execution/instrumentation/ChainedInstrumentation.java:79]: FYI this is a breaking change for eg, a downstream ExecutionStrategy implementation that calls `beginExecution(parameters)` and needs to be changed to call `beginExecution(parameters, parametersInstrumentationState)`.  May want to update the 19.0 release notes which claim no breaking changes.

Comment (andimarek) [reply src/main/java/graphql/execution/instrumentation/ChainedInstrumentation.java:79]: thanks a lot @josephlbarnett ... will update the release notes

**Pull request merged**
